Create draft for Unconference report (#53)
linebender/linebender.github.io: content/blog/2024-05-22-rustl-2024-unconference.md

## The format

The "Unconference" was two days of loosely-structured chats between members of various Rust project teams where people talked about ecosystem collaboration, complained to Lang Team members that their favourite feature wasn't implemented yet, [went to pester everyone else about variadic generics](https://poignardazur.github.io/2024/05/25/report-on-rustnl-variadics/), and other good fun.

Attendees were split into three teams:

- The Rust project.
- GUI and Applications.
- Embedded development.

Discussions inside of these teams was mostly self-organized (more on that later), with cross-team discussion mostly taking place during lunch and recesses.

## The GUI team

As a member of the Linebender project, I was mostly exposed to the GUI team's proceedings.

The main topics of interest were:

- Build tooling (especially for Android and other non-desktop platforms).
- Text layout and edition.
- Winit adoption.
- Accessibility.
- Potential for further collaboration.
- Requests for new Rust features.

Note that I'm skipping a lot topics on the agenda, and other side discussions.
It feels like those topics were the most prominent during those two days.

> There was also a fair amount of informal chatter around LogLog Games' incendiary [Leaving Rust gamedev after 3 years](https://loglog.games/blog/leaving-rust-gamedev/) article.

Reviewer: ```suggestion
There was also a fair amount of informal chatter around LogLog Games' [Leaving Rust gamedev after 3 years](https://loglog.games/blog/leaving-rust-gamedev/) article.
```

I think this term is too strong without a lot of reason to be.

Author: Incendiary is the wrong word, but that article is brazenly critical.

I think those discussions mostly reflected the Bevy community's consensus, that the article had some flaws but overall pointed at very clear weaknesses in the Rust GUI / game engine ecosystem.

### Build tooling

Everyone agreed that a better solution was needed to ship non-TUI Rust apps.

There was discussion about various platforms and build system, and the difficulty of having to eg write Java code for an Android port, and whether anyone wanted to take responsibility for maintaining that glue layer for the rest of the ecosystem.

Unfortunately the discussions on this subject didn't really go anywhere, as far as I could tell.

> Packaging Rust apps is still very much an open problem.

Reviewer: In general? Or just on mobile etc?

Author: To be honest, I dropped out of that discussion pretty quickly. I think someone else than me would be better placed to summarize it.

Reviewer 2: For the Android-specific Java build, I've published the `android-build` crate under Project Robius.

We started some work on packaging Rust apps as of last week, but nothing there to share yet. Unfortunately David, who was working on this, has recently decided to leave the project, so packaging/bundling work has been stalled a bit.

### Text layout and edition

There was some widespread interest in the Linebender ecosystem's text handling crates.
From what I remember, Rik Arends and others had concerns about whether eg Parley and Swash added dependencies to our Vello renderer (they don't).

Some people also wanted Parley to be decoupled from Swash.

People we especially interested in the idea of having access to a common text-editing widget, or at least infrastructure to create one.
That widget would handle IME, text selection, accessibility actions, etc, in a way that would feel native to different platforms with their own text-editing quirks (mobile, MacOS, Linux with vim mode, etc).

### Winit

> By now all of the Rust ecosystem has firmly converged on Winit.

Reviewer: ```suggestion
By now all of the Rust ecosystem has firmly converged on [`winit`](https://github.com/rust-windowing/winit) as the windowing platform-abstraction solution of choice.
```

(Well, not all! [One small project](https://github.com/makepad/makepad) with indomitable maintainers still holds out against the invaders.)

The consensus during this discussion seemed to be that Winit was there to stay, and that any changes we wanted would probably come in the form of upstream contributions, not competing projects.

Glazier is still shelved and we're planning to port its features to Winit until it reaches feature parity.

A discussion that came up was the idea of splitting more "vocabulary types" from Winit into standalone crates.

> Vocabulary crates, eg crates that only define types and very little code, have a very low cost in memory and build types, while allowing projects to interop with eg Winit without actually adding a dependency on Winit.

Reviewer: ```suggestion
Vocabulary crates, eg crates that only define types and very little code, have a very low cost in memory and build times, while allowing projects to interop with eg Winit without actually adding a dependency on Winit.
```

> This was connected with the discussions about text edition: having a vocabulary crate that would list Winit's events and also Parley's cursor movements would allow for a very low-dependency text edition widget.

Reviewer: Text edition?

Author: Text-editing, as in a content-editable textbox.

Reviewer 3: We use 'edition' as the passive form of 'editing' at work, but I don't know if it's actual English. I work for a French company, so it might be a frenchism.

Reviewer: Yeah, that's a Frenchism (as a native English speaker with fluency in French) :)

Author: Huh. It's been a while since I've been caught doing one of those. Usually it's the other way around these days.

Jon Kelley, the creator of Dioxus, had [a laundry list of features](https://dioxus.notion.site/Dioxus-Labs-High-level-Rust-5fe1f1c9c8334815ad488410d948f05e) he wanted from the Rust language. That list was later [filed in the Project Goals repository]((https://github.com/rust-lang/rust-project-goals/pull/10)).

The requests were well-documented, with motivating examples in existing Dioxus code that would clearly be improved by the features.
Most of them felt both self-contained and like they would tremendously help beginners working on their first Rust project.

Overall people were very enthusiastic about them, and the lang team members in the room seemed pretty optimistic that those features could be added.

My favorite proposals were:

- `Capture` trait for auto-cloning reference-counted types in closures.
- Automatic partial borrows.
- Optional struct args.

Jon also wanted support for a remote cache of pre-built crates, so that users wouldn't suffer from having to rebuild dozens of crates every time you download your crate's dependencies.
The consensus from Rust maintainers was that such a cache would be essentially impossible with Rust's current architecture.

> Nico recently released [The borrow checker within](https://smallcultfollowing.com/babysteps/blog/2024/06/02/the-borrow-checker-within/), which feels like a follow-up to some of these discussions.

Reviewer 4: Maybe this should be Niko (Matsakis)?

Reviewer: ```suggestion
Niko Matsakis recently released [The borrow checker within](https://smallcultfollowing.com/babysteps/blog/2024/06/02/the-borrow-checker-within/), which feels like a follow-up to some of these discussions.
```

In-between are self-employed people like me contracted to work on an open-source project, and corporate employees who contribute to the project as part of their 9-to-5 job.

Some monetary contributions can also come from individual non-corporate donors: Servo has about fifty of them, for instance.

With that in mind, some notable sponsors for projects represented at the Unconference were:

- **Google Fonts:** Linebender projects.
- **Futurewei:** Dioxus, Makepad, and Servo.
- **Embark:** Bevy, winit and rust-gpu.
- **Foresight Spatial Labs, MeetKai, Encultured and a few others:** Bevy.
- **Igalia:** Servo.
- **Wyeworks:** Makepad.

Not present at RustNL but relevant to the ecosystem are **System76** (funding iced and COSMIC-Text) and **Slint** who are self-funding as a startup targetting embedded UIs.

Overall the number of different backers feels like a symptom of a healthy ecosystem: while some large corporate sponsors bring much more resources than others (Google and Futurewei especially), the ecosystem isn't in a state where any specific backer pulling out would completely collapse progress.

## Conclusion

> Before anything else, I'd like to address a big thnaks to the volunteers who helped organize RustNL.

Reviewer 5: ```suggestion
Before anything else, I'd like to address a big thanks to the volunteers who helped organize RustNL.
```

Most of the attendees were absolutely exhausted after two days of talks and two more days of Unconference, but these folks had to *run* the darn thing.
I think everyone who attended was fairly impressed.

Overall RustNL and the Unconference felt like getting a year of coordination done in a week.

> It was a thoroughly productive time, and I couldn't be happier I attended. Everyone I asked about it shared the same opinion.

Reviewer 5: ```suggestion
It was a thoroughly productive time, and I couldn't be happier I attended.
Everyone I asked about it shared the same opinion.
```

In the case of the GUI team, people were spread in a very large room, which should have been conducive to small side-discussions and people splitting up to talk about the problems that interested them. Instead, there was an unspoken accord to progress through agenda items one by one, with a few people dominating the discussions on these items.

Because those people we spread around a large room, they had to talk loudly to address each other, which left little room for side-discussions.
Because the people talking were naturally the more confident and extraverted, more introverted people ended up taking a passive role in the discussion.

To me, this feels like a strategic mistake.
It was a setup that encouraged bikeshedding and long back-and-forths, and discouraged plurality of opinions.
It's no coincidence that the most interesting conversations of the Unconference happened at lunch: lunch was the point of the Unconference where people were most mixed, had the most spontaneous conversations, and were least constrained by having to follow what someone else was saying.

> Our discussion were still productive, I just feel like the format could have been improved.

Reviewer 2: thanks for the critiques and feedback here; we'll definitely incorporate this and the other feedback we received for next time. 

Regarding the agenda/topics/structure, given the total absence of replies to my requests before the unconf, I felt like I had no alternative but to bring up the suggested topical agenda at the top of the actual unconf proceedings. Hopefully, with more time to plan ahead of future unconfs (and ideally some additional participation from UI experts in the planning phase), we can receive more input from the attendees and come up with an agreed-upon agenda before we actually meet.

Reviewer 6: The web engines hackfest had people suggest breakout sessions as github issues (https://github.com/Igalia/webengineshackfest/issues). These were then scheduled prior to the start of the conference (which some room for reorganisation) and people could choose which sessions they wanted to attend. I think this is quite a good way of doing it as it allows everyone to participate in negotiating session content rather than putting everything on a central organiser. I think your bound to not get much in the way of responses if people are asked to submit privately.

Not blaming you personally - it's a learning experience for everyone - and you at least stepped up to try and organise things. But some suggestions for next time :)

Reviewer 2: > suggest breakout sessions as github issues

that's a good idea!

> you at least stepped up to try and organise things

i was "voluntold", haha, but yea, someone's gotta do it

- Build tooling (especially for Android and other non-desktop platforms).
- Text layout and editing.
- Winit adoption.
- Accessibility.
- Potential for further collaboration.
- Requests for new Rust features.

Note that I'm skipping a lot topics on the agenda, and other side discussions.
It feels like those topics were the most prominent during those two days.

There was also a fair amount of informal chatter around LogLog Games' strongly critical [Leaving Rust gamedev after 3 years](https://loglog.games/blog/leaving-rust-gamedev/) article.
I think those discussions mostly reflected the Bevy community's consensus, that the article had some flaws but overall pointed at very clear weaknesses in the Rust GUI / game engine ecosystem.

### Build tooling

Everyone agreed that a better solution was needed to ship non-TUI Rust apps.

There was discussion about various platforms and build system, and the difficulty of having to eg write Java code for an Android port, and whether anyone wanted to take responsibility for maintaining that glue layer for the rest of the ecosystem.

Unfortunately the discussions on this subject didn't really go anywhere, as far as I could tell.

> Packaging Rust apps is still very much an open problem.

Reviewer 2: We've confirmed several times over the years that the Cargo team considers this functionality out of scope for Cargo and does not wish to add it. Note that I agree with this, especially since some packaging tools can only be executed on certain platforms (which defeats the promise of cargo's cross-platform nature), so _true_ multi-target builds that cover all platforms of interest aren't even possible.

### Text layout and editing

There was some widespread interest in the Linebender ecosystem's text handling crates.
From what I remember, Rik Arends and others had concerns about whether eg Parley and Swash depend on our Vello renderer (they don't).

Some people also wanted Parley to be decoupled from Swash.

People we especially interested in the idea of having access to a common text-editing widget, or at least infrastructure to create one.
That widget would handle IME, text selection, accessibility actions, etc, in a way that would feel native to different platforms with their own text-editing quirks (mobile, MacOS, Linux with vim mode, etc).

### Winit

By now all of the Rust ecosystem has firmly converged on [`winit`](https://github.com/rust-windowing/winit) as the windowing platform-abstraction solution of choice.

(Well, not all! [One small project](https://github.com/makepad/makepad) with indomitable maintainers still holds out against the invaders.)

> The consensus during this discussion seemed to be that Winit was there to stay, and that any changes we wanted would probably come in the form of upstream contributions, not competing projects.

Reviewer 7: ```suggestion
The consensus during this discussion seemed to be that Winit was there to stay.
[TAO](https://github.com/tauri-apps/tao) users present were generally interested in being able to use upstream winit in the future, if its API adopts or supersedes the improvements made in TAO.
Most were happy to see winit becoming more trait-oriented, and to see the `dpi` crate separated from winit. 
```

Glazier is still shelved and we're planning to port its features to Winit until it reaches feature parity.

A discussion that came up was the idea of splitting more "vocabulary types" from Winit into standalone crates.
Vocabulary crates, eg crates that only define types and very little code, have a very low cost in memory and build times, while allowing projects to interop with eg Winit without actually adding a dependency on Winit.

This was connected with the discussions about text editing: having a vocabulary crate that would list Winit's events and also Parley's cursor movements would allow for a very low-dependency text editing widget.

### Accessibility

*Note: This section is also my personal opinion and doesn't represent the Linebender project either.*

The subject of Accessibility came up informally, but I'll claim it was underexplored.

While the subject was on the agenda, by the time we reached it the Unconference had been running long enough that everyone was tired and ready to leave or get to the catering.

> This is an unfortunate pattern that I think is worth pointing out: if projects don't make a deliberate effort to prioritize discussions of accessibility, the subject will be left out.

Reviewer 2: For the sake of posterity, I'd like to point out that I did attempt to raise the topic publicly twice on Friday, since I wanted to share the results of our discussion with Matt in the office hours from Thursday evening; however, I was unable to do so (and also did not want to forcibly take over room focus) due to seemingly-productive ongoing private discussions that were already naturally occurring.

Note that I feel like this is sort of a damned-if-you-do, damned-if-you-don't situation. What you're asking for here seems to directly contradict the prior complaint about and the desire for a less-centralized discussion flow in favor of multiple smaller groups.

Author: > Note that I feel like this is sort of a damned-if-you-do, damned-if-you-don't situation.

That's not how I meant it, but it's a fair interpretation, and I need to make the text clearer.

To be clear, my position is that organizers are responsible for the IMO overly centralized format, but they're not responsible for the lack of discussion about accessibility. For the latter, I'd say the responsibility is collective, *precisely* because people expect it to be something brought to their attention and not something they have to think about spontaneously.

- [winit](https://github.com/rust-windowing/winit) for creating windows.
- [AccessKit](https://github.com/AccessKit/accesskit/) for plugging into accessibility APIs.
- [Parley](https://github.com/linebender/parley) for rendering text and as a basis for a text-editing widget.

Jon Kelley also said he would be interested in using Masonry as the native backend for Dioxus, though we haven't hashed out details yet.

### Corporate funding

For open-source projects, "Who funds this" is a difficult question: in any healthy project, there is a large scale of contributions, with individual non-corporate contributors at one end, and companies paying cash to the project's treasury at the other.
In-between are self-employed people like me contracted to work on an open-source project, and corporate employees who contribute to the project as part of their 9-to-5 job.

Some monetary contributions can also come from individual non-corporate donors: Servo has about fifty of them, for instance.

With that in mind, some notable sponsors for projects represented at the Unconference were:

- **Google Fonts:** Linebender projects.

> - **Futurewei:** Dioxus, Makepad, and Servo.

Reviewer 2: Kindly add Robius here, which is the primary funding objective from Futurewei.

Note that Futurewei also funds WyeWorks' work on Makepad; not sure if that's relevant.

- **Embark:** Bevy, winit and rust-gpu.

> - **Foresight Spatial Labs, MeetKai, Encultured and a few others:** Bevy.

Reviewer: ```suggestion
- **Foresight Spatial Labs, and a large number of small and medium donors:** Bevy.
```

See https://bevyengine.org/donate/ for a more up to date list. Foresight currently dwarfs other contributions, and MeetKai and Encultured are no longer active donors.

Author: Yeah, I was concerned that listing only big-name donors might offend others. I'll defer to your advice.
